src/main/java/com/cloudbees/hudson/plugins/folder/properties/FolderCredentialsProvider.java

```diff
@@ -522,6 +522,14 @@ private synchronized boolean removeCredentials(@NonNull Domain domain, @NonNull
         private synchronized boolean updateCredentials(@NonNull Domain domain, @NonNull Credentials current,
                                                        @NonNull Credentials replacement) throws IOException {
             checkPermission(CredentialsProvider.UPDATE);
+
+            // See IdCredentials.Helper#equals.
+            // Note: We probably should not assume that all credentials are IdCredentials. We also do this a few lines below in list.indexOf(current) call.
+            // If this one breaks, that one breaks as well.
+            if (!current.equals(replacement)) {
+                throw new IllegalArgumentException("Credentials' IDs do not match, will not update.");
```

Comment:
no no and more no.  you *must* handle credentials that are not IDCredentials

Reply:
What about `int index = list.indexOf(current)` 6 lines below though? It will rely on `equals` under the hood, won't it?

Reply:
Yes it does.
But why do you think that matters?

Reply:
If credential isn't an `IdCredential`, would that `indexOf` correctly find a credential to be updated?

Reply:
> If credential isn't an IdCredential, would that indexOf correctly find a credential to be updated?

Under what conditions would it fail to find `current` in a list of `Credentials`?

Reply:
`Object.equals`.
I've tested locally with the loadfocus plugin, and updating indeed works there. I'm surprised that object reference is that same. I don't know why I assumed it wouldn't be.